petastorm/namedtuple.py

```diff
@@ -0,0 +1,202 @@
+from operator import itemgetter as _itemgetter
+from collections import OrderedDict
+
+class {typename}(tuple):
+    '{typename}({arg_list})'
+
+    __slots__ = ()
+
+    _fields = {field_names!r}
+
+    def __new__(_cls, *args, **kwargs):
+        'Create new instance of {typename}({arg_list})'
+        values = []
+        missings = []
+        for index, field in enumerate(_cls._fields):
+            if index < len(args):
+                values.append(args[index])
+                if field in kwargs:
+                    raise TypeError('__new__() got multiple values for argument %r' % field)
+            elif field in kwargs:
+                values.append(kwargs[field])
+            else:
+                values.append(None)
+                missings.append(field)
+
+        count_missings = len(missings)
+        if count_missings > 0:
+            last_missing = ' and %s' % missings[-1] if count_missings > 1 else ''
+            missings = missings[:-1] if count_missings > 1 else missings
+            raise TypeError('__new__() missing %d required positional argument%s: %s%s' %
+                            (count_missings, 's' if count_missings > 1 else '',
+                            ', '.join(missings), last_missing))
+
+        return _tuple.__new__(_cls, values)
+
+    @classmethod
+    def _make(cls, iterable, new=tuple.__new__, len=len):
+        'Make a new {typename} object from a sequence or iterable'
+        result = new(cls, iterable)
+        if len(result) != {num_fields:d}:
+            raise TypeError('Expected {num_fields:d} arguments, got %d' % len(result))
+        return result
+
+    def _replace(_self, **kwds):
+        'Return a new {typename} object replacing specified fields with new values'
+        result = _self._make(map(kwds.pop, {field_names!r}, _self))
+        if kwds:
+            raise ValueError('Got unexpected field names: %r' % list(kwds))
+        return result
+
+    def __repr__(self):
+        'Return a nicely formatted representation string'
+        return self.__class__.__name__ + '({repr_fmt})' % self
+
+    def _asdict(self):
+        'Return a new OrderedDict which maps field names to their values.'
+        return OrderedDict(zip(self._fields, self))
+
+    def __getnewargs__(self):
+        'Return self as a plain tuple. Used by copy and pickle.'
+        return tuple(self)
+
+{field_defs}
+"""
+
+    _repr_template = '{name}=%r'
+
+    _field_template = '''\
+    {name} = _property(_itemgetter({index}), doc='Alias for field number {index}')
+'''
+
+    __cls = {}
+
+    def _restore_gt_255_fields(name, fields, value):
+        """ Restore an object of namedtuple"""
+        k = (name, fields)
+        cls = __cls.get(k)
+        if cls is None:
+            cls = namedtuple_gt_255_fields(name, fields)
+            __cls[k] = cls
+        return cls(*value)
+
+    def _hack_namedtuple_gt_255_fields(cls):
+        """ Make class generated by namedtuple picklable """
```

Comment:
Can you please document this pickling hack?

Reply:
Ok